File: internal/tmux/session.go
package tmux

import (
	"fmt"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"syscall"
	"time"

	"github.com/bartosz-skejcik/tmux-setup/internal/config"
	"github.com/bartosz-skejcik/tmux-setup/internal/hooks"
)

// CheckDependencies verifies all required dependencies are available
func CheckDependencies(dependencies []string) {
	for _, dep := range dependencies {
		if _, err := exec.LookPath(dep); err != nil {
			log.Fatalf("Dependency missing: %s", dep)
		}
	}
}

// CreateSession creates a new tmux session with the given configuration
func CreateSession(sessionName string, cfg config.Config) error {
	exec.Command("tmux", "new-session", "-d", "-s", sessionName, "-n", "placeholder").Run()
	exec.Command("tmux", "set-option", "-g", "base-index", "1").Run()
	exec.Command("tmux", "set-window-option", "-g", "pane-base-index", "1").Run()

	for i, window := range cfg.Windows {
		if err := createWindow(sessionName, i, window, cfg.Defaults); err != nil {
			return fmt.Errorf("failed to create window %d: %v", i+1, err)
		}
	}

	return nil
}

// AttachSession attaches to an existing tmux session
func AttachSession(sessionName string, focusWindow int) error {
	if focusWindow == 0 {
		focusWindow = 1
	}

	tmuxPath, err := exec.LookPath("tmux")
	if err != nil {
		return err
	}

	exec.Command("tmux", "select-window", "-t", fmt.Sprintf("%s:%d", sessionName, focusWindow)).Run()

	return syscall.Exec(tmuxPath, []string{"tmux", "attach-session", "-t", sessionName}, os.Environ())
}

func createWindow(sessionName string, index int, window config.WindowConfig, defaults config.GlobalDefaults) error {
	windowName := window.Name
	if windowName == "" {
		windowName = fmt.Sprintf("window-%d", index+1)
	}

	if err := hooks.RunPreWindowHooks(window); err != nil {
		return err
	}

	if index == 0 {
		exec.Command("tmux", "rename-window", "-t", fmt.Sprintf("%s:1", sessionName), windowName).Run()
	} else {
		exec.Command("tmux", "new-window", "-t", fmt.Sprintf("%s:%d", sessionName, index+1), "-n", windowName).Run()
	}

	// Set working directory
	dir := resolveDirectory(defaults.Directory, window.Directory)
	if dir != "" {
		sendKeys(sessionName, index+1, 1, fmt.Sprintf("cd %s", dir))
	}

	// Handle Git integration
	if window.GitBranch != "" {
		sendKeys(sessionName, index+1, 1, fmt.Sprintf("git checkout %s", window.GitBranch))
	}

	// Create panes and set up layouts
	if err := createPanes(sessionName, index+1, window, dir); err != nil {
		return err
	}

	if err := hooks.RunPostWindowHooks(window); err != nil {
		return err
	}

	return nil
}

func createPanes(sessionName string, windowIndex int, window config.WindowConfig, defaultDir string) error {
	for i, pane := range window.Panes {
		if i > 0 {
			splitType := "-h"
			if layout, ok := window.Layout.(string); ok && strings.Contains(layout, "vertical") {
				splitType = "-v"
			}
			exec.Command("tmux", "split-window", splitType, "-t", fmt.Sprintf("%s:%d", sessionName, windowIndex)).Run()
		}

		paneDir := resolveDirectory(defaultDir, pane.Directory)
		if paneDir != "" {
			sendKeys(sessionName, windowIndex, i+1, fmt.Sprintf("cd %s", paneDir))
		}

		if pane.InitialCommand != "" {
			sendKeys(sessionName, windowIndex, i+1, pane.InitialCommand)
		}

		if pane.RefreshInterval > 0 {
			go refreshPane(sessionName, windowIndex, i+1, pane)
		}
	}

	// Apply layout
	if window.Layout != nil {
		applyLayout(sessionName, windowIndex, window.Layout)
	}

	return nil
}

func refreshPane(sessionName string, windowIndex, paneIndex int, pane config.PaneConfig) {
	ticker := time.NewTicker(time.Duration(pane.RefreshInterval) * time.Second)
	for range ticker.C {
		sendKeys(sessionName, windowIndex, paneIndex, pane.InitialCommand)
	}
}

func applyLayout(sessionName string, windowIndex int, layout interface{}) {
	switch l := layout.(type) {
	case string:
		exec.Command("tmux", "select-layout", "-t", fmt.Sprintf("%s:%d", sessionName, windowIndex), l).Run()
	case config.LayoutConfig:
		// Apply custom layout using resize-pane commands
		for i, pane := range l.Panes {
			if pane.Width != "" {
				exec.Command("tmux", "resize-pane", "-t", fmt.Sprintf("%s:%d.%d", sessionName, windowIndex, i+1),
					"-x", strings.TrimSuffix(pane.Width, "%")).Run()
			}
			if pane.Height != "" {
				exec.Command("tmux", "resize-pane", "-t", fmt.Sprintf("%s:%d.%d", sessionName, windowIndex, i+1),
					"-y", strings.TrimSuffix(pane.Height, "%")).Run()
			}
		}
	}
}

func sendKeys(sessionName string, windowIndex, paneIndex int, command string) {
	exec.Command("tmux", "send-keys", "-t", fmt.Sprintf("%s:%d.%d", sessionName, windowIndex, paneIndex),
		command, "C-m").Run()
}

func resolveDirectory(parent, child string) string {
	if child != "" {
		if filepath.IsAbs(child) {
			return child
		}
		return filepath.Join(parent, child)
	}
	return parent
}

File: src/lib/session.rs
use anyhow::Error;

use super::{
    config::{Defaults, Layout, Pane, Window},
    hooks::{
        look_path, resolve_dir, run_command, run_post_window_hooks, run_pre_window_hooks, send_keys,
    },
};

pub fn check_deps(deps: &Option<Vec<String>>) -> Result<(), Error> {
    if let Some(deps) = deps {
        for dep in deps.iter() {
            if look_path(dep).is_none() {
                return Err(Error::msg(format!("Dependency not found: {}", dep)));
            }
        }
    }

    Ok(())
}

pub fn create_session(sess_name: &str, config: &crate::Config) -> Result<(), Error> {
    let mut cmd = format!("tmux new-session -d -s {} -n placeholder", sess_name);
    run_command(&cmd)?;

    cmd = "tmux set-option -g base-index 1".to_string();
    run_command(&cmd)?;

    cmd = "tmux set-option -g pane-base-index 1".to_string();
    run_command(&cmd)?;

    config.windows.iter().enumerate().for_each(|(i, window)| {
        match create_window(sess_name, i, window, &config.defaults) {
            Ok(_) => (),
            Err(e) => eprintln!("Error creating window: {}", e),
        }
    });

    Ok(())
}

pub fn create_window(
    sess_name: &str,
    i: usize,
    window: &Window,
    defaults: &Option<Defaults>,
) -> Result<(), Error> {
    let mut cmd = String::new();
    let mut window_name = window.name.clone();

    if window_name.is_empty() {
        window_name = format!("window-{}", i + 1);
    }

    run_pre_window_hooks(window)?;

    if i == 0 {
        cmd = format!("tmux rename-window -t {}:1 {}", sess_name, window_name);
        run_command(&cmd)?;
    } else {
        cmd = format!("tmux new-window -t {sess_name}:{} -n {window_name}", i + 1);
        run_command(&cmd)?;
    }

    let mut resolved_dir = String::new();
    // set working dir
    if let Some(defaults) = defaults {
        if let Some(defaults_directory) = &defaults.directory {
            if let Some(window_directory) = &window.directory {
                resolved_dir = resolve_dir(defaults_directory, &window_directory.to_string());
                send_keys(sess_name, i + 1, 1, &format!("cd {}", resolved_dir))?;
            }
        }
    }

    // handle git branch
    if let Some(git_branch) = &window.git_branch {
        send_keys(sess_name, i + 1, 1, &format!("git checkout {}", git_branch))?;
    }

    // create panes
    // createPanes(sessionName, index+1, window, dir)
    create_panes(sess_name, i + 1, window, resolved_dir)?;

    // run post window hooks
    run_post_window_hooks(window)?;

    Ok(())
}

pub fn create_panes(
    sess_name: &str,
    window_idx: usize,
    window: &Window,
    default_dir: String,
) -> Result<(), Error> {
    // check if the window.panes (Option<Vec<Pane>>) is empty
    if let Some(panes) = &window.panes {
        let mut split_type = "-h";
        for (i, pane) in panes.iter().enumerate() {
            if i > 0 {
                // if the layout is present in the window and the layout has "vertical" set
                // split_type to "-v"
                if let Some(layout) = &window.layout {
                    if layout.to_string().contains("vertical") {
                        split_type = "-v";
                    }
                }
                let cmd = format!("tmux split-window {split_type} -t {sess_name}:{window_idx}");
                run_command(&cmd)?;
            }

            if let Some(pane_dir) = &pane.directory {
                let resolved_dir = resolve_dir(&default_dir, pane_dir);
                send_keys(
                    sess_name,
                    window_idx,
                    i + 1,
                    &format!("cd {}", resolved_dir),
                )?;
            }

            send_keys(sess_name, window_idx, i + 1, &pane.initial_command)?;
        }
    }

    if let Some(window_layout) = &window.layout {
        let cmd = format!(
            "tmux select-layout -t {sess_name}:{window_idx} {layout}",
            sess_name = sess_name,
            window_idx = window_idx,
            layout = window_layout
        );
        run_command(&cmd)?;
    }

    Ok(())
}
